src/engine/Union.cpp

```diff
@@ -347,172 +346,250 @@ struct Wrapper {
     std::shared_ptr<const Result> result2) const {
   using Range = std::variant<Result::LazyResult, std::array<Wrapper, 1>>;
-  Range leftRange = result1->isFullyMaterialized()
-                        ? Range{std::array{Wrapper{result1->idTable(),
-                                                   result1->localVocab()}}}
-                        : Range{std::move(result1->idTables())};
-  Range rightRange = result2->isFullyMaterialized()
-                         ? Range{std::array{Wrapper{result2->idTable(),
-                                                    result2->localVocab()}}}
-                         : Range{std::move(result2->idTables())};
-
-  // Clang falsely reports that the this capture is unused here.
-  DISABLE_UNUSED_LAMBDA_CAPTURE_WARNINGS
+  auto toRange = [](const auto& result) {
+    return result->isFullyMaterialized()
+               ? Range{std::array{
+                     Wrapper{result->idTable(), result->localVocab()}}}
+               : Range{std::move(result->idTables())};
+  };
+  Range leftRange = toRange(result1);
+  Range rightRange = toRange(result2);
+
+  auto end = ql::ranges::find_if(targetOrder_, [this](ColumnIndex index) {
+    const auto& [left, right] = _columnOrigins.at(index);
+    return left == NO_COLUMN || right == NO_COLUMN;
+  });
+  std::span trimmedTargetOrder{targetOrder_.begin(),
+                               end == targetOrder_.end() ? end : end + 1};
+
+  auto applyPermutation = [this](IdTable idTable,
+                                 const std::vector<ColumnIndex>& permutation) {
+    return transformToCorrectColumnFormat(std::move(idTable), permutation);
+  };
+
   return std::visit(
-      [this, requestLaziness, &result1, &result2](auto leftRange,
-                                                  auto rightRange) {
+      [this, requestLaziness, &result1, &result2, &trimmedTargetOrder,
+       &applyPermutation]<typename LR, typename RR>(
+          LR leftRange, RR rightRange) -> Result::LazyResult {
         return ad_utility::callFixedSize(
-            targetOrder_.size(),
-            [this, requestLaziness, &result1, &result2, &leftRange,
-             &rightRange]<int COMPARATOR_WIDTH>() {
-              return computeResultKeepOrderImpl<COMPARATOR_WIDTH>(
-                  requestLaziness, std::move(leftRange), std::move(rightRange),
-                  std::pair{std::move(result1), std::move(result2)});
+            trimmedTargetOrder.size(),
+            [this, requestLaziness, &result1, &result2, &leftRange, &rightRange,
+             &trimmedTargetOrder, &applyPermutation]<int COMPARATOR_WIDTH>() {
+              constexpr size_t extent = COMPARATOR_WIDTH == 0
+                                            ? std::dynamic_extent
+                                            : COMPARATOR_WIDTH;
+              return Result::LazyResult{
+                  SortedUnionImpl<extent, LR, RR, decltype(applyPermutation)>(
```

Comment:
Maybe with a deduction guide you can get rid of some of the explicit template parameters.